packages/editor/src/components/block-removal-warnings/index.js

```diff
@@ -0,0 +1,52 @@
+/**
+ * WordPress dependencies
+ */
+
+import { __ } from '@wordpress/i18n';
+import { privateApis as blockEditorPrivateApis } from '@wordpress/block-editor';
+
+/**
+ * Internal dependencies
+ */
+import { unlock } from '../../lock-unlock';
+
+const { BlockRemovalWarningModal } = unlock( blockEditorPrivateApis );
+
+// Prevent accidental removal of certain blocks, asking the user for confirmation first.
+const blockRemovalRules = {
+	'core/query': {
+		message: __(
+			'The Query Loop block displays a list of posts or pages, so removing it will prevent that content from displaying.'
+		),
+	},
+	'core/post-content': {
+		message: __(
+			'Removing the Post Content block will stop your post or page content from displaying on this template.'
+		),
+	},
+	'core/post-template': {
+		message: __(
+			'The Post Template block displays each post or page in a Query Loop, so removing it will stop post content displaying in your query loop.'
+		),
+	},
```

Comment:
A good reason to introduce a callback is to have some consistency in how the system works, but here there's still some blocks working one way (no callback) and then pattern overrides working another way (a callback).

Another problem with how it works is that some object keys are meaningful and represent one thing—`core/query`, `core/post-content` etc. are block names and are used for matching, while `core/pattern-overrides` is a block binding source and the key doesn't do anything.

As its all private code there's a chance to be a bit bolder, take a step back and consider how it all might work ideally to meet all the requirements.

Reply:
I was working from the point of view that it is reasonably standard for an API to have optional parameters and:
```js
const blockName = select.getBlockName( clientId );
if ( rules[ blockName ] ) {
	blockNamesForPrompt.add( blockName );
}
```
is already very generic so need point in passing it with every rule, the `callBack` can just be optional for rules that require something less generic.

But, there is no panic to get this change in, so we might as well take some time to think about the best approach.

Reply:
one option would be to separate block specific and generic rules and handle each group separately in the selector, eg.

```js
const blockSpecific = {
	'core/query': {
		message: __(
			'The Query Loop block displays a list of posts or pages, so removing it will prevent that content from displaying.'
		),
	},
	'core/post-content': {
		message: __(
			'Removing the Post Content block will stop your post or page content from displaying on this template.'
		),
	},
	'core/post-template': {
		message: __(
			'The Post Template block displays each post or page in a Query Loop, so removing it will stop post content displaying in your query loop.'
		),
	},
};
const generic = {
	'core/pattern-overrides': {
		message: __(
			'Deleting a block with pattern instance overrides can break other blocks on your site that have content linked to it.'
		),
		callBack: ( select, clientId, blockNamesForPrompt ) => {
			const blockAttributes = select.getBlockAttributes( clientId );
			if (
				blockAttributes?.metadata?.bindings &&
				JSON.stringify( blockAttributes.metadata.bindings ).includes(
					'core/pattern-overrides'
				)
			) {
				blockNamesForPrompt.add( 'core/pattern-overrides' );
			}
		},
	},
};

const blockRemovalRules = { blockSpecific, generic };
```

Reply:
will have another think about various approaches/naming, etc. tomorrow.